src/internal.ts

```diff
@@ -0,0 +1,138 @@
+import { createHmac } from 'crypto'
+import { request as httpsRequest } from 'https'
+import { stringify } from 'querystring'
+
+import {
+  API_HOST,
+  API_PATH,
+  API_PROTOCOL,
+  API_VERSION,
+  API_FORMAT,
+  API_VALID_RESPONSE,
+} from './constants'
+import CoinpaymentsError from './error'
+import { validatePayload } from './validation'
+
+import {
+  CoinpaymentsCredentials,
+  CoinpaymentsRequest,
+  CoinpaymentsInternalRequestOps,
+  CoinpaymentsInternalResponse,
+  CoinpaymentsReturnCallback,
+} from './types/base'
+
+export const getPrivateHeaders = (
+  credentials: CoinpaymentsCredentials,
+  options: CoinpaymentsRequest
+) => {
+  const { secret } = credentials
+
+  const paramString = stringify(options)
+  const signature = createHmac('sha512', secret)
+    .update(paramString)
+    .digest('hex')
+
+  return {
+    'Content-Type': 'application/x-www-form-urlencoded',
+    HMAC: signature,
+  }
+}
+
+export const resolveRequest = async <ExpectedResponse>(
+  reqOps: CoinpaymentsInternalRequestOps,
+  options: CoinpaymentsRequest,
+  callback?: CoinpaymentsReturnCallback<ExpectedResponse>
+): Promise<ExpectedResponse> => {
+  let response
+  try {
+    response = await makeRequest<ExpectedResponse>(reqOps, options)
+  } catch (e) {
+    if (callback) {
+      return callback(e)
+    }
+    return Promise.reject(e)
+  }
+  if (callback) {
+    return callback(null, response)
+  }
+  return Promise.resolve(response)
```

Comment:
Since this is an async function, everything it returns is wrapped in a promise anyway, no need to do that explicitly.